lib/firenest/topology/erlang.ex

```diff
@@ -133,112 +98,327 @@ defmodule Firenest.Topology.Erlang do
+    # Node names are also stored in an ETS table for fast lookups.
+    persist_node_names(topology, %{})
+
+    # We generate a unique ID to be used alongside the node name
+    # to guarantee uniqueness in case of restarts. Then we do a
+    # broadcast over the Erlang topology to find other processes
+    # like ours. The other server monitor states will be carried
+    # in their pongs.
     id = id()
-    Enum.each(Node.list(), &ping(&1, topology, id))
-    {:ok, %{topology: topology, nodes: %{}, id: id, subscribers: %{}}}
+    Enum.each(Node.list(), &ping(&1, topology, id, []))
+
+    state = %{
+      topology: topology,
+      nodes: %{},
+      id: id,
+      monitors: %{},
+      local_names: %{},
+      subscribers: %{}
+    }
+
+    {:ok, state}
   end
 
+  ## Local messages
+
   def handle_call({:subscribe, pid}, _from, state) do
     ref = Process.monitor(pid)
     state = put_in(state.subscribers[ref], pid)
     {:reply, ref, state}
   end
 
-  def handle_call({:unsubscribe, ref}, _from, state) do
-    {_, state} = pop_in(state.subscribers[ref])
-    {:noreply, :ok, state}
+  # Receives the sync monitor command from a local process and broadcast
+  # this monitor is up in all known instances of this topology.
+  def handle_call({:sync_named, pid, name}, _from, state) do
+    %{topology: topology, id: id, nodes: nodes} = state
+
+    case maybe_remove_dead_monitor(state, name) do
+      {:ok, state} ->
+        ref = Process.monitor(name)
+        state = put_in(state.monitors[ref], name)
+        state = put_in(state.local_names[name], {pid, ref})
+
+        nodes =
+          for {node, {id, _, remote_names}} <- nodes,
+              Map.has_key?(remote_names, name),
+              do: {node, id}
+
+        topology_broadcast(topology, {:monitor_up, Kernel.node(), id, name, ref})
+        {:reply, {:ok, nodes}, state}
+
+      {:error, existing_pid} ->
+        {:reply, {:error, {:already_synced, existing_pid}}, state}
+    end
+  end
+
+  def handle_info({:DOWN, ref, _, pid, _}, state) when Kernel.node(pid) != Kernel.node() do
+    {:noreply, delete_node(state, pid, ref)}
+  end
+
+  def handle_info({:DOWN, ref, _, _, _}, %{monitors: monitors} = state) do
+    case monitors do
+      %{^ref => _} ->
+        {:noreply, remove_dead_monitor(state, ref)}
+
+      %{} ->
+        {_, state} = pop_in(state.subscribers[ref])
+        {:noreply, state}
+    end
   end
 
-  @doc false
-  def handle_info({:nodeup, node, _}, %{topology: topology, id: id} = state) do
-    ping(node, topology, id)
+  ## Distributed messages
+
+  # This is the message received from remote nodes when they have a
+  # local monitor up.
+  def handle_info({:monitor_up, node, id, name, monitor_ref}, state) do
+    %{nodes: nodes, local_names: local_names} = state
+
+    state =
+      case nodes do
+        # We know this node. The other node guarantees to deliver a monitor_down
+        # before monitor_up for the same name, so we don't need to check this here.
```

Comment:
This guarantee does not hold because while there is order guarantee there is neither delivery guarantee nor connection knowledge. It is possible for a `pong` to arrive that was caused by a previous `ping`, or a `ping` triggered by a prior `nodeup`, which causes a new process monitor. As delivery isn't guaranteed the next message to arrive could be a `monitor_up` for a name in the monitors of `ping`/`pong`. We can solve this by using lamport clocks.

Reply:
Isn't the `id` serving a similar purpose?

Reply:
The `id` is unique for the lifetime of a pid on a node which does not cover this situation.

Reply:
In terms of lamport clocks the `id` would be good to identify an entity but not any logical time.

Reply:
> This guarantee does not hold because while there is order guarantee there is neither delivery guarantee nor connection knowledge.

However, the message is has guaranteed delivery unless we see a nodedown/DOWN message from that node, which would cause this to be removed. Or am I missing something?